csharp/src/Ice/Identity.cs

```diff
@@ -11,81 +12,104 @@ namespace ZeroC.Ice
-            }
 
-            string category;
-            string? name = null;
-            if (slash == -1)
-            {
-                try
-                {
-                    name = StringUtil.UnescapeString(s, 0, s.Length, "/");
-                }
-                catch (ArgumentException ex)
-                {
-                    throw new FormatException($"invalid name in identity `{s}", ex);
-                }
-                category = "";
-            }
-            else
-            {
-                try
+                string category;
+                string? name = null;
+                if (slash == -1)
                 {
-                    category = StringUtil.UnescapeString(s, 0, slash, "/");
-                }
-                catch (ArgumentException ex)
-                {
-                    throw new FormatException($"invalid category in identity `{s}", ex);
+                    try
+                    {
+                        name = StringUtil.UnescapeString(s, 0, s.Length, "/");
+                    }
+                    catch (ArgumentException ex)
+                    {
+                        throw new FormatException($"invalid name in identity `{s}'", ex);
+                    }
+                    category = "";
                 }
-
-                if (slash + 1 < s.Length)
+                else
                 {
                     try
                     {
-                        name = StringUtil.UnescapeString(s, slash + 1, s.Length, "/");
+                        category = StringUtil.UnescapeString(s, 0, slash, "/");
                     }
                     catch (ArgumentException ex)
                     {
-                        throw new FormatException($"invalid name in identity `{s}", ex);
+                        throw new FormatException($"invalid category in identity `{s}'", ex);
+                    }
+
+                    if (slash + 1 < s.Length)
+                    {
+                        try
+                        {
+                            name = StringUtil.UnescapeString(s, slash + 1, s.Length, "/");
+                        }
+                        catch (ArgumentException ex)
+                        {
+                            throw new FormatException($"invalid name in identity `{s}'", ex);
+                        }
                     }
                 }
-            }
 
-            if (string.IsNullOrEmpty(name))
-            {
-                throw new FormatException($"empty name in identity `{s}'");
+                return name?.Length > 0 ? new Identity(name, category) :
+                    throw new FormatException($"invalid empty name in identity `{s}'");
             }
+            else
+            {
+                string name;
+                string? category = null;
 
-            return new Identity(name, category);
+                string[] segments = s.Split('/');
+                switch (segments.Length) // always > 0
```

Comment:
How do we know it's always > 0? Should this be an assert then or should we fix the error message in case it's 0?

Reply:
If you split an empty string, you get one empty segment. I don't think we need an assert for Split's behavior.